### PR TITLE
[release-4.13] WINC-1043: Test upgrading to CSI storage, from in-tree drivers

### PR DESCRIPTION
**Pull request by sebsoto**
Manual cherry-pick of https://github.com/openshift/windows-machine-config-operator/pull/1705

### CODE REVIEW

File: test/e2e/delete_test.go [hunk 1]
```diff
@@ -270,3 +271,20 @@ func (tc *testContext) waitUntilDaemonsetScaled(name string, desiredReplicas int
 	}
 	return nil
 }
+
+// cleanupWorkloads attempts to delete all deployments that exist within the testContext workload namespace
+func (tc *testContext) cleanupDeployments() {
+	deployments, err := tc.client.K8s.AppsV1().Deployments(tc.workloadNamespace).List(context.TODO(), meta.ListOptions{})
```
Comment (aravindhp) [inline test/e2e/delete_test.go:277]: Do we need to wait for the deployments to be deleted? PS: I think I might have missed reviewing the master PR.

File: test/e2e/upgrade_test.go [hunk 2]
```diff
@@ -216,6 +217,17 @@ func TestUpgrade(t *testing.T) {
...
 
+	if inTreeUpgrade {
+		// Deploy the CSI drivers and wait for a CSI node to be created
+		// This is a requirement for storage workloads to go back to ready
+		azureProvider, ok := tc.CloudProvider.(*azure.Provider)
+		require.True(t, ok, "in tree upgrade must be ran on Azure")
+		require.NoError(t, azureProvider.EnsureWindowsCSIDaemonSet(tc.client.K8s))
+		log.Printf("waiting for csinodes to reflect driver deployment")
+		err = tc.waitForCSINodesWithDrivers(gc.allNodes())
+		require.NoError(t, err)
```
Comment (aravindhp) [inline test/e2e/upgrade_test.go:228]: This can be a helper method

**Review by aravindhp — APPROVED** [review]

File: test/e2e/upgrade_test.go [hunk 3]
```diff
@@ -241,3 +253,32 @@ func (tc *testContext) testWorkloadsAvailable(t *testing.T) {
 		})
 	assert.NoError(t, err)
 }
+
+// waitForCSINodes waits for a CSINode to exist for each node with a driver loaded, indicating CSI functionality has
+// been enabled for the given node.
+func (tc *testContext) waitForCSINodesWithDrivers(nodes []v1.Node) error {
+	return wait.PollImmediate(retry.Interval, 10*time.Minute, func() (bool, error) {
```
Comment (jrvaldes) [inline test/e2e/upgrade_test.go:260]: I feel [retry.Timeout ](https://github.com/openshift/windows-machine-config-operator/blob/833c1b6b6c24c02de9ac68e7f903d092b8b889aa/pkg/retry/retry.go#L13) may be a good fit here.

Comment (sebsoto) [reply test/e2e/upgrade_test.go:260]: agreed, should have done that in the original PR

**Pull request merged**